### ISSUE TITLE
db hint

### ISSUE THREAD

**Issue by vitaly-t** [sazzer/wbjs#1]
There is no need for the use of `pg` module for anything that you do ;) in fact, you should remove it as a dependency ;)

**Comment by sazzer** [comment]
How come?

I'm using it directly as - I believe - I need it for the postgres-cleaner
module to work correctly. All of the actual database code that I'm doing
myself is currently using pg-promise because I feel it gives me a cleaner,
easier interface whilst still letting me write the queries how I want to do
so - especially in how it can understand if a query is meant to return 0, 1
or many rows and make it easier to access the data. I'm also going to look
at a SQL Query Builder for dynamic queries soon - possibly Knex or Squel -
rather than hard-coding the queries myself. (I know you can use Knex to
actually execute queries as well, but it looks like the mechanism to run
raw queries that I've written myself is very much an afterthought and the
main emphasis is query building)

**Comment by vitaly-t** [comment]
`pg` module is contained within `pg-promise`, see: https://github.com/vitaly-t/pg-promise/wiki/FAQ#how-to-access-the-instance-of-node-postgres-thats-used

**Comment by sazzer** [comment]
Fair enough - I didn't realise that. I'll have a play and see if it works
for me :)

btw - when I sent the previous message saying my preference for pg-promise,
I'd not actually realised you were the author of it! I'm very impressed
with it so thank you :)

**Comment by vitaly-t** [comment]
>I'm also going to look
at a SQL Query Builder for dynamic queries soon - possibly Knex or Squel -
rather than hard-coding the queries myself. (I know you can use Knex to
actually execute queries as well, but it looks like the mechanism to run
raw queries that I've written myself is very much an afterthought and the
main emphasis is query building)

The choice is - universality (Knex) versus ultimate flexibility (pg-promise). If variable between development for PostgreSQL and other databases, knex may help you. But purely for postgreSQL, pg-promise has the upper hand ;)

**Comment by sazzer** [comment]
It's purely Postgres for now, though I always try to stick to ANSI SQL
where possible just in case.

The benefit of a SQL Query builder for me is less about database
independence and more about having cleaner code for dynamic queries.
Especially when there are an arbitrary number of clauses such as in a
search. It's the difference between string concatenation and needing to get
the brackets right and all that, or a fluent api that does it all for me...

I'm also going to look
at a SQL Query Builder for dynamic queries soon - possibly Knex or Squel -
rather than hard-coding the queries myself. (I know you can use Knex to
actually execute queries as well, but it looks like the mechanism to run
raw queries that I've written myself is very much an afterthought and the
main emphasis is query building)

The choice is - universality (Knex) versus ultimate flexibility
(pg-promise). If variable between development for PostgreSQL and other
databases, knex may help you. But purely for postgreSQL, pg-promise has the
upper hand ;)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/sazzer/wbjs/issues/1#issuecomment-224988278>, or mute
the thread
<https://github.com/notifications/unsubscribe/AAECGNaSHfKVGPJcph2k3jjJVW1do86Bks5qKF5JgaJpZM4Ix_cl>
.

**Comment by vitaly-t** [comment]
I saw some projects where people would connect Knex to format queries, and then use pg-promise to execute them. That's a possibility ;)

**Comment by sazzer** [comment]
Exactly what I was thinking :) not 100% sure on knex or squel or which yet,
but that's the idea :)

**Comment by vitaly-t** [comment]
also, you do not need to write code like this:
```js
    return new Promise((resolve, reject) => {
      pgp.pg.connect(dbUrl, (err, client, done) => {
        if (err) {
          done();
          reject(err);
        } else {
          resolve({client, done});
        }
      })
```
you have [db.connect](http://vitaly-t.github.io/pg-promise/Database.html#.connect) that's way better, and safer.

**Comment by vitaly-t** [comment]
https://github.com/sazzer/wbjs/blob/master/src/server/worlds/dao.js

u must never use `.catch` before `.then` with promises, because you will get into `.then` after getting into `.catch`
